src/main.rs

```diff
@@ -1,3 +1,829 @@
+use serde::de::Visitor;
+use serde::{de, de::Deserialize, ser::Serialize};
+use serde::{Deserializer, Serializer};
+use simplelog::{
+    ColorChoice, ConfigBuilder, Level, LevelFilter, TermLogger, TerminalMode, WriteLogger,
+};
+
+type Version = cargo_metadata::Version;
+
+#[derive(Parser)]
+#[clap(version, about, long_about = None)]
+#[clap(propagate_version = true)]
+struct Cli {
+    /// Subcommands ("no subcommand" is its own subcommand)
+    #[clap(subcommand)]
+    command: Option<Commands>,
+
+    // Cargo flags we support and forward to e.g. 'cargo metadata'
+    #[clap(flatten)]
+    manifest: clap_cargo::Manifest,
+    #[clap(flatten)]
+    workspace: clap_cargo::Workspace,
+    #[clap(flatten)]
+    features: clap_cargo::Features,
+
+    // Top-level flags
+    /// Do not pull in new "audits".
+    #[clap(long)]
+    locked: bool,
+
+    /// How verbose logging should be (log level).
+    #[clap(short, long, arg_enum)]
+    #[clap(default_value_t = Verbose::Warn)]
+    verbose: Verbose,
+
+    /// Instead of stdout, write output to this file.
+    #[clap(long)]
+    output_file: Option<PathBuf>,
+
+    /// Instead of stderr, write logs to this file (only used after successful CLI parsing).
+    #[clap(long)]
+    log_file: Option<PathBuf>,
+}
+
+#[derive(Subcommand)]
+enum Commands {
+    /// initialize cargo-vet for your project
+    Init(InitArgs),
+    /// Fetch the source of `$crate $version`
+    Fetch(FetchArgs),
+    /// Yield a diff against the last reviewed version.
+    Diff(DiffArgs),
+    /// Mark `$crate $version` as reviewed with `$message`
+    Certify(CertifyArgs),
+    /// Mark `$crate $version` as unacceptable with `$message`
+    Forbid(ForbidArgs),
+    /// Suggest some low-hanging fruit to review
+    Suggest(SuggestArgs),
+    /// ??? List audits mechanisms ???
+    Audits(AuditsArgs),
+    /// Print --help as markdown (for generating docs)
+    HelpMarkdown(HelpMarkdownArgs),
+}
+
+#[derive(clap::Args)]
+struct InitArgs {}
+
+#[derive(clap::Args)]
+struct FetchArgs {
+    krate: String,
+    version: String,
+}
+
+#[derive(clap::Args)]
+struct DiffArgs {}
+
+#[derive(clap::Args)]
+struct CertifyArgs {
+    krate: String,
+    version: String,
+    message: String,
+}
+
+#[derive(clap::Args)]
+struct ForbidArgs {
+    krate: String,
+    version: String,
+    message: String,
+}
+
+#[derive(clap::Args)]
+struct SuggestArgs {}
+
+#[derive(clap::Args)]
+struct AuditsArgs {}
+
+#[derive(clap::Args)]
+struct HelpMarkdownArgs {}
+
+/// Logging verbosity levels
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, ArgEnum)]
+enum Verbose {
+    Off,
+    Error,
+    Warn,
+    Info,
+    Debug,
+    Trace,
+}
+
+#[derive(serde::Deserialize)]
+struct Config {
+    // Reserved for future use, if not present version=1 assumed.
+    // (not sure whether this versions the format, or semantics, or...
+    // for now assuming this species global semantics of some kind.
+    version: Option<u64>,
```

Comment:
I don't really know what this is "for" but it felt right to toss it in.

Reply:
Sure